pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarPrimitiveSchemaHandler.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.sql.presto;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Date;
+
+import io.netty.util.concurrent.FastThreadLocal;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.impl.schema.AutoConsumeSchema;
+import org.apache.pulsar.common.schema.SchemaInfo;
+
+/**
+ * A presto schema handler that interprets data using pulsar schema.
+ */
+public class PulsarPrimitiveSchemaHandler implements SchemaHandler {
+
+    private final SchemaInfo schemaInfo;
+    private final Schema<?> schema;
+
+    public PulsarPrimitiveSchemaHandler(SchemaInfo schemaInfo) {
+        this.schemaInfo = schemaInfo;
+        this.schema = AutoConsumeSchema.getSchema(schemaInfo);
+    }
+
+    @Override
+    public Object deserialize(ByteBuf byteBuf) {
+        byte[] data = ByteBufUtil.getBytes(byteBuf);
+        Object currentRecord = schema.decode(data);
+        switch (schemaInfo.getType()) {
```

Comment:
for these date, time, and timestamp types, I think we can just decode with the LongSchema.  For example there is no point in deserializing to a "Date" object just to get the long value via "getTime()"

Reply:
in principle I would suggest not shortcutting the schema implementation. it would make maintenance become very hard. because if we can anything in schema implementation, we might forget updating the part in presto. For example, if we change the implementation in Date schema to include tz information. In presto we are using LongSchema which bypass the whole backward compatibility logic, which will cause problems.

I would actually encourage sticking to using Schema and let Schema take care of all the bc handling. If this part become a bottleneck, we can always seek an approach to improve it. Correctness and maintenance is the top priority for pulsar-presto at this moment.

Reply:
@sijie schema in Pulsar is created for serializing from a POJO to bytes and vice versa.  This is a different scenario then what happens in the presto connector especially here.  What is different about presto connector is that we need to map types to presto types.  Which is different goal then what schema is designed for.

>  For example, if we change the implementation in Date schema to include tz information

We will have to change the logic here regardless if we decided to add tz info.  The current logic still returns a long (timestamp) at the end of the day which will not contain any TZ info.  It does by creating unnecessary object like "Date". 

>  Correctness and maintenance is the top priority for pulsar-presto at this moment.

Oh I didn't know we have set priorities for pulsar-presto.  Was this discussed somewhere?

Reply:
Also if we are intending to use only existing schemas for deserializing data in the presto connector,  I would suggest optimizing the schemas first for use cases in the presto connector e.g. adding support for ByteBuf, use DSL JSON, etc, before we replace current deserializing code.

Reply:
@jerrypeng 

> This is a different scenario then what happens in the presto connector especially here. What is different about presto connector is that we need to map types to presto types. Which is different goal then what schema is designed for.

How is that different? The schema is for defining the types stored in Pulsar. It is the source of truth for interpreting the Pulsar types. Because it handles schema versioning and schema evaluation. It is not just simply deserialization and serialization. 

If Presto wants to deserialize and serialize the raw bytes, it can define its way to serialize and deserialize the raw bytes. But if it is using the schema type defined in Pulsar, it should sticky to Pulsar schema. Because Pulsar schema handles versioning and evaluation. It is technically wrong to bypass it.

> We will have to change the logic here regardless if we decided to add tz info. The current logic in this PR still returns a long (timestamp) at the end of the day which will not contain any TZ info. It does by creating unnecessary object like "Date".

Yes. That's called schema evaluation. That's why we have added versioning and many other features to handle schema evaluation. And that's why I am so strong on using Schema interface and implementation. You can't bypass it. At the time you bypass the schema interface and implementation, you are bypassing schema versioning and schema evaluation. When a schema evolved, you have to re-implement the logic in presto. That's is going to be a disaster.

Reply:
> Also if we are intending to use only existing schemas for deserializing data in the presto connector, I would suggest optimizing the schemas first for use cases in the presto connector e.g. adding support for ByteBuf, use DSL JSON, etc, before we replace current deserializing code.

@jerrypeng this PR is NOT replacing any deserializing code. It doesn't touch AVRO and JSON. This PR is adding the support for new types. It doesn't change any existing behaviors. I have created a follow up issue to optimize the date types. 

We can optimize schemas first before changing AVRO and JSON to use schema implementation. But it doesn't make any sense to  block this PR.

Reply:
> @jerrypeng this PR is NOT replacing any deserializing code. It doesn't touch AVRO and JSON. This PR is adding the support for new types. It doesn't change any existing behaviors. I have created a follow up issue to optimize the date types.

We can optimize schemas first before changing AVRO and JSON to use schema implementation. But it doesn't make any sense to block this PR.

My comment is about the future work not about this PR.  Sorry if that wasn't clear. In an ideal situation we should just use schema for deserializing data in presto, but I don't want to sacrifice performance to do so

Reply:
> In an ideal situation we should just use schema for deserializing data in presto, but I don't want to sacrifice performance to do so

I think we agreed on this.